src/cdi_ndi/NDI_CP_Eloss.cc

```diff
@@ -0,0 +1,288 @@
+
+  int num_temperatures = 0;
+  ndi_error = NDI2_get_int_val(dataset_handle, NDI_NUM_TEMPS, &num_temperatures);
+  Require(ndi_error == 0);
+  n_temperature = static_cast<uint32_t>(num_temperatures);
+
+  temperatures.resize(n_temperature);
+  ndi_error = NDI2_get_float64_vec(dataset_handle, NDI_TEMPS, temperatures.data(),
+                                   static_cast<int>(temperatures.size()));
+  Require(ndi_error == 0);
+  min_log_temperature = temperatures.front();
+  d_log_temperature = temperatures[1] - temperatures[0];
+  min_temperature = exp(min_log_temperature);
+  max_temperature = exp(min_log_temperature + d_log_temperature * n_temperature);
+
+  stopping_data_1d.resize(n_energy * n_density * n_temperature);
+  ndi_error = NDI2_get_float64_vec_x(
+      dataset_handle, NDI_TARGET_DEDX, std::to_string(target.get_zaid()).c_str(),
+      stopping_data_1d.data(), static_cast<int>(stopping_data_1d.size()));
+  Require(ndi_error == 0);
+
+  // Check for uniform log spacing
+  for (uint32_t n = 1; n < n_energy; n++) {
+    Require(rtt_dsxx::soft_equiv(d_log_energy, energies[n] - energies[n - 1], 1.e-5));
+  }
+  for (uint32_t n = 1; n < n_density; n++) {
+    Require(rtt_dsxx::soft_equiv(d_log_density, densities[n] - densities[n - 1], 1.e-5));
+  }
+  for (uint32_t n = 1; n < n_temperature; n++) {
+    Require(rtt_dsxx::soft_equiv(d_log_temperature, temperatures[n] - temperatures[n - 1], 1.e-5));
+  }
+
+  // Convert units on table to match those of getEloss:
+  //   energy:      MeV -> cm/shk (using target particle mass)
+  double energy_cgs = exp(min_log_energy) * (1.e6 * pc.electronVolt());
+  min_log_energy = log(sqrt(2. * energy_cgs / target.get_mass()) * 1.e-8);
+  d_log_energy = d_log_energy / 2.;
+  //   density:     cm^-3 -> g cm^-3
+  min_log_density = log(exp(min_log_density) * target.get_mass());
+  //   temperature: keV -> keV
+  // Note that d log x = dx / x is not affected by unit conversion factors
+  for (auto &energy : energies) {
+    energy = sqrt(2. * (energy * 1.e6 * pc.electronVolt()) / target.get_mass()) * 1.e-8;
+  }
+  for (auto &density : densities) {
+    density *= target.get_mass();
+  }
+}
+//----------------------------------------------------------------------------//
+/*!
+ * \brief Interpolate the tabulated stopping power for a given material and
+ *        projectile state.
+ * \param[in] temperature Material temperature [keV]
+ * \param[in] density Material density [g cm^-3]
+ * \param[in] partSpeed Particle speed [cm shk^-1]
+ */
+double NDI_CP_Eloss::getEloss(const double temperature, const double density,
+                              const double partSpeed) const {
+  if (temperature <= min_temperature || temperature >= max_temperature || density <= min_density ||
+      density >= max_density || partSpeed <= min_energy || partSpeed >= max_energy) {
+    // Outside of the table
+    return 0.;
+  }
+  //                              return 0.;;
```

Comment:
I think maybe this line can go...

Reply:
just a little stylistic flourish... I will remove though :(